### PR TITLE
feat(trace viewer): add "Copy as Playwright Request" button

### LINKED ISSUE #33258: [Feature]: Add ability to "Copy as Playwright Code" from traceviewer
### 🚀 Feature Request

As the team has recently added the functionality to copy http requests as cURL and as fetch, it would be great to have the option to copy the the http request to playwright code using the "request". This would allow someone to quickly post the playwright code into a playwright test or into a separate file to help build out a test data factory quickly. The big timesaver here is getting the headers and body created from the request.

### Example

After recording a playwright test, viewing an html report traceviewer, or after a playwright run while in ui mode, I could visit the network traffic tab, and right click a specific request and select copy as Playwright Code. 

### Motivation

This would help me build out helper functions to enable the creation test data.

### PR DESCRIPTION
**Pull request by Skn0tt**
Closes https://github.com/microsoft/playwright/issues/33258. See the added test for some code examples.

### CODE REVIEW

File: packages/playwright-core/src/utils/isomorphic/codegen.ts [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import type * as har from '@trace/har';
+
+export function generatePlaywrightRequestCall(request: har.Request, body: string | undefined): string {
```
Comment (pavelfeldman) [inline packages/playwright-core/src/utils/isomorphic/codegen.ts:19]: Sounds like we are using this only from the web, so put it into packages/trace-viewer/src ?

File: tests/playwright-test/codegen.spec.ts [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+    method: 'GET',
+    headers: [{ name: 'User-Agent', value: 'Mozilla/5.0' }, { name: 'Date', value: '2021-01-01' }],
+    httpVersion: '1.1',
+    cookies: [],
+    queryString: [],
+    headersSize: 0,
+    bodySize: 0,
+    comment: '',
+  }, 'foo')).toEqual(`
+await page.request.get('http://example.com/foo', {
```
Comment (pavelfeldman) [inline tests/playwright-test/codegen.spec.ts:32]: What about the other languages? Also, most of your code is around formatting, please add more tests to cover edge condition. This test belongs to `tests/library/unit` as it does not use e2e capabilities too.

Comment (Skn0tt) [reply tests/playwright-test/codegen.spec.ts:32]: > What about the other languages?

I wanted to keep things simple so I only started with Javascript. We can extend it when users ask for it.

**Review by Skn0tt — COMMENTED** [review]
Will move the files around + add some tests for edge cases 👍

File: packages/trace-viewer/src/ui/workbench.tsx [hunk 3]
```diff
@@ -225,7 +225,7 @@ export const Workbench: React.FunctionComponent<{
     id: 'network',
     title: 'Network',
     count: networkModel.resources.length,
-    render: () => <NetworkTab boundaries={boundaries} networkModel={networkModel} onEntryHovered={setHighlightedEntry}/>
+    render: () => <NetworkTab boundaries={boundaries} networkModel={networkModel} onEntryHovered={setHighlightedEntry} sdkLanguage={model?.sdkLanguage ?? 'javascript'} />
```
Comment (Skn0tt) [inline packages/trace-viewer/src/ui/workbench.tsx:228]: i've added a fallback here to make types easier to reason about. this is only unavailable before the model is loaded, so falling back to `javascript` in a loading state is alright I feel.

**Comment by Skn0tt** [comment]
I've updated the PR with codegen logic for all supported languages.

File: packages/trace-viewer/src/ui/codegen.ts [hunk 4]
```diff
@@ -0,0 +1,287 @@
...
+    // Handle null and undefined
+    if (obj === null)
+      return 'null';
+    if (obj === undefined)
+      return 'undefined';
+
+    // Handle primitive types
+    if (typeof obj !== 'object') {
+      if (typeof obj === 'string')
+        return `'${obj}'`;
```
Comment (pavelfeldman) [inline packages/trace-viewer/src/ui/codegen.ts:58]: This does not look like a proper string escaping. Same problem across the languages where you only escape quotes and do not escape \\.

File: packages/trace-viewer/src/ui/codegen.ts [hunk 5]
```diff
@@ -0,0 +1,287 @@
...
+    // Handle null and undefined
+    if (obj === null)
+      return 'None';
+    if (obj === undefined)
+      return 'None';
+
+    // Handle primitive types
+    if (typeof obj !== 'object') {
+      if (typeof obj === 'string')
+        return `"${obj.replaceAll('"', '\\"')}"`;
```
[the remark this answers is not in the document]
Comment (Skn0tt) [reply packages/trace-viewer/src/ui/codegen.ts]: done in https://github.com/microsoft/playwright/pull/33298/commits/a798374d1238edff59134eaa294cbf803db96abf

File: packages/trace-viewer/src/ui/codegen.ts (hunk 4 above)
Comment (Skn0tt) [reply packages/trace-viewer/src/ui/codegen.ts]: good catch! fixed in https://github.com/microsoft/playwright/pull/33298/commits/a798374d1238edff59134eaa294cbf803db96abf

File: packages/trace-viewer/src/ui/codegen.ts [hunk 6]
```diff
@@ -0,0 +1,287 @@
...
+
+    // Handle regular objects
+    if (Object.keys(obj).length === 0)
+      return 'new {}';
+    const spaces = ' '.repeat(level * indent);
+    const nextSpaces = ' '.repeat((level + 1) * indent);
+
+    const entries = Object.entries(obj).map(([key, value]) => {
+      const formattedValue = this.prettyPrintObject(value, indent, level + 1);
+      const formattedKey = level === 0 ? key : `["${key}"]`;
```
[the remark this answers is not in the document]
Comment (Skn0tt) [reply packages/trace-viewer/src/ui/codegen.ts]: the keys in level 0 are only `Method`, `Params` etc. All other levels getting the same string literal escape treatment now.

**Review by Skn0tt — COMMENTED** [review]
addressed!

File: packages/trace-viewer/src/ui/codegen.ts [hunk 7]
```diff
@@ -0,0 +1,305 @@
...
+        key :
+        this.stringLiteral(key);
+      return `${nextSpaces}${formattedKey}: ${formattedValue}`;
+    }).join(',\n');
+
+    return `{\n${entries}\n${spaces}}`;
+  }
+
+  private stringLiteral(v: string): string {
+    v = v.replace(/\\/g, '\\\\').replace(/'/g, '\\\'');
```
Comment (pavelfeldman) [inline packages/trace-viewer/src/ui/codegen.ts:95]: This does not escape control characters like \n, \r, \t. `JSON.stringify(v)` would do a better job.

File: packages/trace-viewer/src/ui/codegen.ts [hunk 8]
```diff
@@ -0,0 +1,305 @@
...
+      options.headers = Object.fromEntries(request.headers.map(header => [header.name, header.value]));
+
+    const params = [`'${urlParam}'`];
+    const hasOptions = Object.keys(options).length > 0;
+    if (hasOptions)
+      params.push(this.prettyPrintObject(options));
+    return `await page.request.${method}(${params.join(', ')});`;
+  }
+
+  private prettyPrintObject(obj: any, indent = 2, level = 0): string {
```
Comment (pavelfeldman) [inline packages/trace-viewer/src/ui/codegen.ts:48]: For JS, I wonder if JSON.stringify would do a similar job.

File: packages/trace-viewer/src/ui/codegen.ts [hunk 9]
```diff
@@ -0,0 +1,305 @@
...
+    const entries = Object.entries(obj).map(([key, value]) => {
+      const formattedValue = this.prettyPrintObject(value, indent, level + 1);
+      return `${nextSpaces}${this.stringLiteral(key)}: ${formattedValue}`;
+    }).join(',\n');
+
+    return `{\n${entries}\n${spaces}}`;
+  }
+
+  private stringLiteral(v: string): string {
+    v = v.replace(/\\/g, '\\\\').replace(/"/g, '\\"');
```
Comment (pavelfeldman) [inline packages/trace-viewer/src/ui/codegen.ts:173]: Same here, maybe `repr` function could be used

File: packages/trace-viewer/src/ui/codegen.ts [hunk 10]
```diff
@@ -0,0 +1,305 @@
...
+      const formattedKey = level === 0 ? key : `[${this.stringLiteral(key)}]`;
+      return `${nextSpaces}${formattedKey} = ${formattedValue}`;
+    }).join(',\n');
+
+    return `new() {\n${entries}\n${spaces}}`;
+  }
+
+  private stringLiteral(v: string): string {
+    // escape douvle quotes and backslashes
+    v = v.replace(/\\/g, '\\\\').replace(/"/g, '\\"');
```
Comment (pavelfeldman) [inline packages/trace-viewer/src/ui/codegen.ts:258]:
```suggestion
    v = v.replace(/\\/g, '\\\\').replace(/"/g, '\\"').replace(/\n/g, '\\n').replace(/\r/g, '\\r').replace(/\t/g, '\\t');
  return $"\"{v}\"";
```

File: packages/trace-viewer/src/ui/codegen.ts [hunk 11]
```diff
@@ -0,0 +1,305 @@
...
+      options.push(`setHeader(${this.stringLiteral(header.name)}, ${this.stringLiteral(header.value)})`);
+
+    if (options.length > 0)
+      params.push(`RequestOptions.create()\n  .${options.join('\n  .')}\n`);
+    return `request.${method}(${params.join(', ')});`;
+  }
+
+  private stringLiteral(v: string): string {
+    // escape douvle quotes and backslashes
+    v = v.replace(/\\/g, '\\\\').replace(/"/g, '\\"');
```
Comment (pavelfeldman) [inline packages/trace-viewer/src/ui/codegen.ts:290]:
```suggestion
    v = v.replace(/\\/g, '\\\\').replace(/"/g, '\\"').replace(/\n/g, '\\n').replace(/\r/g, '\\r').replace(/\t/g, '\\t');  
```

File: packages/trace-viewer/src/ui/codegen.ts (hunk 7 above)
Comment (Skn0tt) [reply packages/trace-viewer/src/ui/codegen.ts:95]: The problem with `JSON.stringify` is that it outputs double-quotes, and I want the outputted code to conform to the style we have in the rest of the product, which uses single quotes.

> This does not escape control characters like \n, \r, \t.

Good catch! I'll address that.

File: packages/trace-viewer/src/ui/codegen.ts (hunk 9 above)
Comment (Skn0tt) [reply packages/trace-viewer/src/ui/codegen.ts]: Are you referring to pythons `__repr__`? I don't think we can use that in here :/

File: packages/trace-viewer/src/ui/codegen.ts (hunk 8 above)
Comment (Skn0tt) [reply packages/trace-viewer/src/ui/codegen.ts:48]: It would certainly output correct JS and be much less brittle, but it would contain lots of pesky double quotes that don't match the code style we have in the rest of the product.

File: packages/trace-viewer/src/ui/codegen.ts (hunk 10 above)
Comment (Skn0tt) [reply packages/trace-viewer/src/ui/codegen.ts]: Now that I think about it, I think this one could use `JSON.stringify`!

File: packages/trace-viewer/src/ui/codegen.ts (hunk 11 above)
Comment (Skn0tt) [reply packages/trace-viewer/src/ui/codegen.ts]: Same here, I think this one could also use JSON.stringify

File: packages/trace-viewer/src/ui/codegen.ts (hunk 9 above)
Comment (Skn0tt) [reply packages/trace-viewer/src/ui/codegen.ts]: I've replaced it with `JSON.stringify`, I think that should work.

File: packages/trace-viewer/src/ui/codegen.ts (hunk 7 above)
Comment (Skn0tt) [reply packages/trace-viewer/src/ui/codegen.ts:95]: I've made it detect control characters and output them unescaped into source - I think the resulting code is easier to reason about than having a long string of escaped control characters.

**Review by Skn0tt — COMMENTED** [review]
addressed

**Review by pavelfeldman — APPROVED** [review]
